adam-core/src/main/scala/org/bdgenomics/adam/rdd/GenomicRDD.scala

```diff
@@ -119,6 +118,480 @@ trait GenomicRDD[T, U <: GenomicRDD[T, U]] {
+          //of regions as there are T's in this partition.
+          while (List(listRepresentation(i)._1(j), previousReferenceRegion).sorted.head != previousReferenceRegion) {
+            j += 1
+          }
+          listBuffer += ((listRepresentation(i)._1(j), listRepresentation(i)._2))
+        }
+      }
+      listBuffer.toIterator
+    }
+
+    /**
+     * This is a function that splits each partition (iter) into lists keyed with the destination
+     * partition number. It preserves the order of the data received and thus is good for moving
+     * around sorted data. Each partition is given the same amount of data.
+     *
+     * @param iter The partition data. (k,v) where k is the index in the RDD and v is T
+     * @param average the average number of tuples on all partitions
+     * @return grouped lists keyed with the destination partition number. These lists maintain
+     *         their original order
+     */
+    def getBalancedPartitionNumber(iter: Iterator[(Long, T)], partitionMap: (ReferenceRegion, ReferenceRegion), average: Double): Iterator[(Int, ((ReferenceRegion, ReferenceRegion), List[(Long, T)]))] = {
+      // converting to list so we can package data that is going to the same node with a groupBy
+      val correctPartition = inferCorrectReferenceRegionsForPartitionWithZip(iter, partitionMap)
+      correctPartition.map(f => ((f._1 / average).toInt, f)).toList.groupBy(_._1).mapValues(f => f.map(_._2))
+        .map(f => (f._1, ((f._2.head._2._1, f._2.last._2._1), f._2.map(g => (g._1, g._2._2))))).toIterator
+    }
+
+    /**
+     * The purpose of this method is to co-partition two RDDs according to their ReferenceRegions.
+     * It is useful during ShuffleRegionJoins because it guarantees no extra shuffling to get correct
+     * and complete region joins. RDD is already sorted coming in, so we take precaution to keep it
+     * that way.
+     *
+     * The simplest way of sending data to the correct node is to just send everything that has a
+     * start position that could fall within the boundaries
+     *
+     * This is best used under the condition that (repeatedly) repartitioning is more expensive than
+     * calculating the proper location of the records of this.rdd. It requires a pass through the
+     * co-located RDD to get the correct partition(s) for each record. It will assign a record to
+     * multiple partitions if necessary.
+     *
+     * @param rddToCoPartitionWith the rdd that this.rdd will be copartitioned with
+     * @return returns the newly repartitioned rdd that is co-partitioned with rddToCoPartitionWith
+     */
+    def coPartitionByGenomicRegion[X, Y <: GenomicRDD[X, Y]](rddToCoPartitionWith: GenomicRDD[X, Y])(implicit tTag: ClassTag[T], xTag: ClassTag[X]): U = {
+      assert(rddToCoPartitionWith.SortedTrait.isSorted)
+      val collectedDestinationPartitionMap = rddToCoPartitionWith.SortedTrait.partitionMap
```

Comment:
Can you add inline docs to this function? Once inline docs are added, I'll make a review pass.